app/core/Tracker2/pageUnload.js

```diff
@@ -0,0 +1,72 @@
+function getKeyForNamespace (namespace) {
+  return `coco.tracker.pageUnloadRetries.${namespace}`;
+}
+
+function getNamespaceData (namespace) {
+  const namespaceKey = getKeyForNamespace(namespace)
+  const namespaceRetriesJson = window.sessionStorage.getItem(namespaceKey)
+
+  return JSON.parse(namespaceRetriesJson || '[]')
+}
+
+function setNamespaceData(namespace, data) {
+  const namespaceKey = getKeyForNamespace(namespace)
+
+  if (data) {
+    window.sessionStorage.setItem(namespaceKey, JSON.stringify(data))
+  } else {
+    window.sessionStorage.removeItem(namespaceKey)
+  }
+}
+
+function runAfterPageLoad (namespace, identifier, args) {
+  const namespaceRetries = getNamespaceData(namespace)
+
+  namespaceRetries.push({
+    identifier,
+    args,
+    at: new Date().getTime()
+  })
+
+  setNamespaceData(namespace, namespaceRetries)
+}
+
+export async function watchForPageUnload (timeout = 500) {
+  let unloadCallback;
+  const unloadPromise = new Promise((resolve, reject) => {
+    unloadCallback = () => reject('unload')
+    window.addEventListener('beforeunload', unloadCallback)
+  })
+
+  const timeoutPromise = new Promise((resolve) => setTimeout(resolve, timeout));
+
+  try {
+    await Promise.race([ unloadPromise, timeoutPromise ])
+  } finally {
+    window.removeEventListener('beforeunload', unloadCallback)
+  }
+}
+
+export async function retryOnPageUnload (namespace, identifier, args, func, timeout = 500) {
```

Comment:
I think a comment for why this function is needed or when it is appropriate to use would be helpful.

This function and this file in general is so wildly cool.

Reply:
Done